### ISSUE TITLE
Add comments to a generated playlist

### ISSUE THREAD

**Issue by CandleCandle** [carlanton/m3u8-parser#8]
RFC 8216 allows for comments to be included in a manifest. They are any line that starts with a `#` and does not then have `EXT` as the following three characters. (section 4.1, paragraph 3)

## Options
1. * (media playlist) Have all comments as a block immediately preceding the media segments.
   * (master playlist) Have all comments as a block immediately preceding the renditions.
2. (both playlists) Attempt to introduce some ordering to the tags, allowing for comments between tags, and around media segments, renditions and variants.

Option (1) looks to be the easiest, however, (2) allows for more flexibility.

**Comment by carlanton** [comment]
Good idea!

I like option 2, but I'm afraid that tag ordering (which is cool) would require a lot of changes to the API to support that (addBefore, addAfter, etc. etc.).

I believe that the main use-case for comments is to describe tags (variants, alternate renditions). A third option could be to preserve the comments to the following tag in the playlist.

The API could look something like this:

```java
MasterPlaylist playlist = MasterPlaylist.builder()
    .version(4)
    .independentSegments(true)
    .comments("top comment", "multi-line thing") // <--
    .addAlternativeRenditions(AlternativeRendition.builder()
        .type(MediaType.AUDIO)
        .name("Default audio")
        .groupId("AUDIO")
        .comments("default audio rendition") // <--
        .build())
    .addVariants(
        Variant.builder()
            .addCodecs("avc1.4d401f", "mp4a.40.2")
            .bandwidth(900000)
            .uri("v0.m3u8")
            .comments("low bandwidth") // <-- 
            .build(),
        Variant.builder()
            .addCodecs("avc1.4d401f", "mp4a.40.2")
            .bandwidth(900000)
            .uri("v1.m3u8")
            .resolution(1280, 720)
            .comments("high bandwidth", "whatever") // <-- 
            .build())
    .build();
```

Result:
```
#EXTM3U
# top comment
# multi-line thing
#EXT-X-VERSION:4
#EXT-X-INDEPENDENT-SEGMENTS
# default audio rendition
#EXT-X-MEDIA:TYPE=AUDIO,GROUP-ID="AUDIO",NAME="Default audio"
# low bandwidth
#EXT-X-STREAM-INF:BANDWIDTH=900000,CODECS="avc1.4d401f,mp4a.40.2"
v0.m3u8
# high bandwidth
# whatever
#EXT-X-STREAM-INF:BANDWIDTH=900000,CODECS="avc1.4d401f,mp4a.40.2",RESOLUTION=1280x720
v1.m3u8
```

It's not as flexible as option 2 but it should be much easier to implement in the current code base. What do you think?

**Comment by carlanton** [comment]
We could also treat comments with only newlines as blank lines (without '#') to improve readability:

```
...
        Variant.builder()
            .addCodecs("avc1.4d401f", "mp4a.40.2")
            .bandwidth(900000)
            .uri("v1.m3u8")
            .resolution(1280, 720)
            .comments("\n", "\n", "whatever") // <-- 
            .build())
...
```

```
#EXTM3U
# top comment
# multi-line thing
#EXT-X-VERSION:4
#EXT-X-INDEPENDENT-SEGMENTS
# default audio rendition
#EXT-X-MEDIA:TYPE=AUDIO,GROUP-ID="AUDIO",NAME="Default audio"
# low bandwidth
#EXT-X-STREAM-INF:BANDWIDTH=900000,CODECS="avc1.4d401f,mp4a.40.2"
v0.m3u8


# whatever
#EXT-X-STREAM-INF:BANDWIDTH=900000,CODECS="avc1.4d401f,mp4a.40.2",RESOLUTION=1280x720
v1.m3u8
```

**Comment by CandleCandle** [comment]
> It's not as flexible as option 2 but it should be much easier to implement in the current code base. What do you think?

That looks like a reasonable compromise. The location of the block of comments *I* would put after the `#EXT-X-VERSION:4` line, but that does not really matter.

> We could also treat comments with only newlines as blank lines (without '#') to improve readability:

Sure. I believe the specification allows for completely empty lines.

> `.comments("\n", "\n", "whatever") // <-- `

However, if the API takes a varargs of comments, then an empty string should result in a line prefixed with a `#`. Does the comments block allow for arbitrary text and therefore potentially allow the library to generate an invalid manifest? - I would suggest that the javadoc contains something like "you are responsible for supplying inputs that do not cause the manifest to be invalid"

**Comment by brendonmoss** [comment]
Is there any WIP around adding comments to this package like what's discussed above?

**Comment by carlanton** [comment]
I'm afraid not. I got lost in parsing ambiguities :) PR:s are always welcome!

**Comment by brendonmoss** [comment]
I'm needing to add some custom tags to the master m3u8, comments seems to be the best possible way to achieve this based on the flexibility in the RFC when not prefixed with EXT.

I am not too interested in maintaining a fork long term so ideally merging back in would be my preference else I can explore some manipulation to the output after building the manifest up and continue to use this package as it is right now.

let me know if you're willing to review merge requests for changes around adding in Option 1 functionality.
Thanks.

**Comment by carlanton** [comment]
I'll gladly review a PR for that :-)

**Comment by carlanton** [comment]
fixed in #35
